census/estimator/trainer/model.py

```diff
@@ -0,0 +1,246 @@
+    an int64 Tensor representing the label index for classification tasks, 
+    and a float32 Tensor representing the value for a regression task.
+  """
+  # Build a Hash Table inside the graph
+  table = tf.contrib.lookup.string_to_index_table_from_tensor(
+      tf.constant(LABELS))
+
+  # Use the hash table to convert string labels to ints
+  return table.lookup(label_string_tensor)
+
+
+# ************************************************************************
+# YOU NEED NOT MODIFY ANYTHING BELOW HERE TO ADAPT THIS MODEL TO YOUR DATA
+# ************************************************************************
+
+
+def column_to_dtype(column):
+  if isinstance(column, layers.feature_column._SparseColumn):
+    return tf.string
+  else:
+    return tf.float32
+
+
+def serving_input_fn():
+    feature_placeholders = {
+        column.name: tf.placeholder(column_to_dtype(column), [None])
+        for column in INPUT_COLUMNS
+    }
+    features = {
+      key: tf.expand_dims(tensor, -1)
+      for key, tensor in feature_placeholders.items()
+    }
+    return input_fn_utils.InputFnOps(
+      features,
+      None,
+      feature_placeholders
+    )
+
+def generate_input_fn(filenames,
+                      num_epochs=None,
+                      shuffle=True,
+                      skip_header_lines=0,
+                      batch_size=40):
+  """Generates an input function for training or evaluation.
+  Args:
+      filenames: [str] list of CSV files to read data from.
+      num_epochs: int how many times through to read the data.
+        If None will loop through data indefinitely
+      shuffle: bool, whether or not to randomize the order of data.
+        Controls randomization of both file order and line order within
+        files.
+      skip_header_lines: int set to non-zero in order to skip header lines
+        in CSV files.
+      batch_size: int First dimension size of the Tensors returned by
+        input_fn
+  Returns:
+      A function () -> (features, indices) where features is a dictionary of
+        Tensors, and indices is a single Tensor of label indices.
+  """
+  def _input_fn():
+    filename_queue = tf.train.string_input_producer(
+        filenames, num_epochs=num_epochs, shuffle=shuffle)
+    reader = tf.TextLineReader(skip_header_lines=skip_header_lines)
+
+    _, rows = reader.read_up_to(filename_queue, num_records=batch_size)
+    
+    # DNNLinearCombinedClassifier expects rank 2 tensors.
+    row_columns = tf.expand_dims(rows, -1)
+    columns = tf.decode_csv(row_columns, record_defaults=CSV_COLUMN_DEFAULTS)
+    features = dict(zip(CSV_COLUMNS, columns))
+    
+    # Remove unused columns
+    used_column_names = [col.name for col in INPUT_COLUMNS]
+    used_column_names.append(LABEL_COLUMN)
+    for col in CSV_COLUMNS:
+      if col not in used_column_names:
+        features.pop(col)
+
+    if shuffle:
+      features = tf.train.shuffle_batch(
```

Comment:
I had a tough time understanding what's going on here. Comments would be helpful